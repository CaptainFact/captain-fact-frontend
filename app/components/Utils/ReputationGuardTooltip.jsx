import React from 'react'
import { withNamespaces, Trans } from 'react-i18next'
import { Link } from 'react-router'
import Popup from 'reactjs-popup'

import { Icon } from './Icon'
import Message from './Message'
import { withLoggedInUser } from '../LoggedInUser/UserProvider'

export const ReputationGuardTooltip = ({
  t,
  checkReputation,
  requiredRep,
  children,
  tooltipPosition = 'bottom center',
}) => {
  return checkReputation(requiredRep) ? (
    children({ hasReputation: true })
  ) : (
    <Popup
      position={tooltipPosition}
      contentStyle={{ zIndex: 999, maxWidth: 250 }}
      on="hover"
      trigger={<div className="help-tooltip-trigger">{children({ hasReputation: false })}</div>}
    >
      <Message type="primary">
        <Icon name="info-circle" />
        &nbsp;
        <Trans i18nKey="errors:client.needReputation" requiredRep={requiredRep}>
          You need at least {{ requiredRep }}
          <Link to="/help/reputation">{t('pages.reputation').toLowerCase()}</Link>
          points to do that
        </Trans>
      </Message>
    </Popup>
  )
}

export default withNamespaces('help')(withLoggedInUser(ReputationGuardTooltip))
