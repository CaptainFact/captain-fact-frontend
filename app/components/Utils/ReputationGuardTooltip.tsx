import React from 'react'
import { Trans, useTranslation } from 'react-i18next'
import { Link } from 'react-router-dom'

import { Tooltip, TooltipContent, TooltipTrigger } from '@/components/ui/tooltip'

import { useLoggedInUser } from '../LoggedInUser/UserProvider'

export const ReputationGuardTooltip = ({
  requiredRep,
  children,
  tooltipPosition = 'bottom',
}: {
  requiredRep: number
  children: (props: { hasReputation: boolean }) => React.ReactNode
  tooltipPosition?: 'bottom' | 'top' | 'right' | 'left'
}) => {
  const { t } = useTranslation('help')
  const { checkReputation } = useLoggedInUser()
  if (checkReputation(requiredRep)) {
    return children({ hasReputation: true })
  }

  return (
    <Tooltip delayDuration={0}>
      <TooltipTrigger asChild>
        <div className="help-tooltip-trigger">{children({ hasReputation: false })}</div>
      </TooltipTrigger>
      <TooltipContent side={tooltipPosition}>
        <Trans i18nKey="errors:client.needReputation" requiredRep={requiredRep}>
          You need at least {{ requiredRep }}
          <Link to="/help/reputation" className="text-white underline">
            {t('pages.reputation').toLowerCase()}
          </Link>
          points to do that
        </Trans>
      </TooltipContent>
    </Tooltip>
  )
}

export default ReputationGuardTooltip
