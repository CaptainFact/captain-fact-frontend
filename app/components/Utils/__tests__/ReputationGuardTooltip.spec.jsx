/**
 * @jest-environment jsdom
 */

import { t } from 'react-i18next'

import { ReputationGuardTooltip } from '../ReputationGuardTooltip'

const DEFAULT_CHILDREN = ({ hasReputation }) => `Has reputation: ${hasReputation}`

test("user doesn't have required reputation", () => {
  snapshotComponent(
    <ReputationGuardTooltip t={t} requiredRep={42} checkReputation={() => false}>
      {DEFAULT_CHILDREN}
    </ReputationGuardTooltip>
  )
})

test('user have required reputation', () => {
  snapshotComponent(
    <ReputationGuardTooltip t={t} requiredRep={42} checkReputation={() => true}>
      {DEFAULT_CHILDREN}
    </ReputationGuardTooltip>
  )
})
