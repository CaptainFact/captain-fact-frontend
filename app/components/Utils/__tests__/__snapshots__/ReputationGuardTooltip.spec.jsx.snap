// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user doesn't have required reputation 1`] = `
<Popup
  arrow={true}
  arrowStyle={
    Object {
      "background": "#f9fbfb",
    }
  }
  className=""
  closeOnDocumentClick={true}
  closeOnEscape={true}
  contentStyle={
    Object {
      "zIndex": 999,
    }
  }
  defaultOpen={false}
  keepTooltipInside={false}
  lockScroll={false}
  modal={false}
  mouseEnterDelay={100}
  mouseLeaveDelay={100}
  offsetX={0}
  offsetY={0}
  on="hover"
  onClose={[Function]}
  onOpen={[Function]}
  open={false}
  overlayStyle={Object {}}
  position="bottom center"
  trigger={
    <div
      className="help-tooltip-trigger"
    >
      Has reputation: false
    </div>
  }
>
  <Message
    type="primary"
  >
    <Icon
      name="info-circle"
    />
    Â 
    <Trans
      i18nKey="errors:client.needReputation"
      requiredRep={42}
    >
      You need at least 
      <Component />
      <Link
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/help/reputation"
      >
        translated[pages.reputation]
      </Link>
      points to do that
    </Trans>
  </Message>
</Popup>
`;

exports[`user have required reputation 1`] = `"Has reputation: true"`;
